internal/connection_manager.go

```diff
@@ -148,55 +177,46 @@ func (cm *connectionManager) getConnectionInternal(address core.Address, asOwner
 	return nil
 }
 
-func (cm *connectionManager) ConnectionCount() int32 {
-	cm.lock.RLock()
-	defer cm.lock.RUnlock()
-	return int32(len(cm.connections))
-}
+// following methods are called under same connectionsMutex writeLock
+// only entry point is getOrCreateConnectionInternal
+func (cm *connectionManagerImpl) getOrCreateConnectionInternal(address core.Address, asOwner bool) (*Connection, error) {
+	cm.connectionsMutex.Lock()
+	defer cm.connectionsMutex.Unlock()
 
-func (cm *connectionManager) getOrConnectInternal(address core.Address, asOwner bool) (*Connection, error) {
-	cm.lock.Lock()
-	defer cm.lock.Unlock()
+	conn, found := cm.connections[address.String()]
 
-	conn := cm.getConnectionInternal(address, asOwner)
-	if conn != nil {
-		return conn, nil
+	if !found {
+		return cm.createConnection(address, asOwner)
 	}
 
-	if !asOwner && cm.client.ClusterService.ownerConnectionAddress.Load().(*protocol.Address).Host() == "" {
-		return nil, core.NewHazelcastIllegalStateError("ownerConnection is not active", nil)
-	}
-	invocationService := cm.client.InvocationService
-	connectionID := cm.NextConnectionID()
-	con := newConnection(address, invocationService.responseChannel, invocationService.notSentMessages, connectionID, cm)
-	if con == nil {
-		return nil, core.NewHazelcastTargetDisconnectedError("target is disconnected", nil)
+	if !asOwner {
+		return conn, nil
 	}
-	err := cm.authenticate(con, asOwner)
 
-	if err != nil {
-		return nil, err
+	if conn.isOwnerConnection {
+		return conn, nil
 	}
+	err := cm.authenticate(conn, asOwner)
 
-	return con, nil
-}
-
-func (cm *connectionManager) getOwnerConnection() *Connection {
-	ownerConnectionAddress := cm.client.ClusterService.ownerConnectionAddress.Load().(*protocol.Address)
-	if ownerConnectionAddress.Host() == "" {
-		return nil
+	if err == nil {
+		return conn, nil
 	}
-	return cm.getActiveConnection(ownerConnectionAddress)
+	return nil, err
+}
 
+func (cm *connectionManagerImpl) NextConnectionID() int64 {
+	cm.nextConnectionID = cm.nextConnectionID + 1
+	return cm.nextConnectionID
 }
 
-func (cm *connectionManager) authenticate(connection *Connection, asOwner bool) error {
+func (cm *connectionManagerImpl) encodeRequest(asOwner bool) *protocol.ClientMessage {
```

Comment:
I think we can name this as `encodeAuthenticationRequest` as this method only encodes Authentication requests.

Reply:
ok